universal-login-contracts/test/contracts/ERC1077MasterCopy.js

```diff
@@ -0,0 +1,314 @@
+      });
+
+      describe('refund', () => {
+        it('should refund in token after execute transfer ethers', async () => {
+          msg = {...transferMessage, from: identityProxy.address, gasToken: mockToken.address};
+          signature = await calculateMessageSignature(privateKey, msg);
+          data = executeSignedFunc.encode([...getExecutionArgs(msg), signature]);
+
+          await wallet.sendTransaction({to: identityProxy.address, data, gasPrice, gasLimit});
+          expect(await mockToken.balanceOf(wallet.address)).to.be.above(relayerTokenBalance);
+        });
+
+        it('should refund after execute transfer ethers', async () => {
+          const transaction = await wallet.sendTransaction({to: identityProxy.address, data, gasPrice, gasLimit});
+          const {gasUsed} = await provider.getTransactionReceipt(transaction.hash);
+          const totalCost = gasUsed.mul(utils.bigNumberify(gasPrice));
+
+          expect(await wallet.getBalance()).to.be.above(relayerBalance.sub(totalCost));
+        });
+      });
+    });
+
+    describe('failed execution of transfer', () => {
+      it('nonce too low', async () => {
+        await wallet.sendTransaction({to: identityProxy.address, data, gasPrice, gasLimit});
+        await expect(wallet.sendTransaction({to: identityProxy.address, data, gasPrice, gasLimit}))
+          .to.be.revertedWith('Invalid nonce');
+      });
+
+      it('nonce too high', async () => {
+        msg = {...transferMessage, from: identityProxy.address, nonce: 2};
+        signature = await calculateMessageSignature(privateKey, msg);
+        data = executeSignedFunc.encode([...getExecutionArgs(msg), signature]);
+
+        await expect(wallet.sendTransaction({to: identityProxy.address, data, gasPrice, gasLimit}))
+          .to.be.revertedWith('Invalid nonce');
+      });
+
+      it('emits ExecutedSigned event', async () => {
+        msg = {...failedTransferMessage, from: identityProxy.address};
+        signature = await calculateMessageSignature(privateKey, msg);
+        data = executeSignedFunc.encode([...getExecutionArgs(msg), signature]);
+        const messageHash = calculateMessageHash(msg);
+
+        await expect(wallet.sendTransaction({to: identityProxy.address, data, gasPrice, gasLimit}))
+          .to.emit(identityMaster, 'ExecutedSigned')
+          .withArgs(messageHash, 0, false);
+      });
+
+      it('should increase nonce, even if transfer fails', async () => {
+        msg = {...failedTransferMessage, from: identityProxy.address};
+        signature = await calculateMessageSignature(privateKey, msg);
+        data = executeSignedFunc.encode([...getExecutionArgs(msg), signature]);
+
+        await wallet.sendTransaction({to: identityProxy.address, data, gasPrice, gasLimit});
+        expect(await provider.getBalance(to)).to.eq(parseEther('0.0'));
+        expect(await proxyAsIdentity.lastNonce()).to.eq(1);
+      });
+
+      describe('Invalid signature', () => {
```

Comment:
We should have two tests here: `no signature` and `invalid signature`.
Each should attempt to execute a transaction.